interpreter/exec/eval.ml

```diff
@@ -737,6 +742,154 @@ let rec step (c : config) : config =
+            Refer (Aggr.(ArrayRef (Array (td, fsd))));
+            Plain (Const (I32 (I32.add dst 1l) @@ e.at));
+            Refer (Aggr.(ArrayRef (Array (ts, fss))));
+            Plain (Const (I32 (I32.add src 1l) @@ e.at));
+            Plain (Const (I32 (I32.sub sz 1l) @@ e.at));
+            Plain (ArrayCopy (x, y));
+          ]
+        else (* d > s *)
+          vs', List.map (Lib.Fun.flip (@@) e.at) [
+            Refer (Aggr.(ArrayRef (Array (td, fsd))));
+            Plain (Const (I32 (I32.add dst 1l) @@ e.at));
+            Refer (Aggr.(ArrayRef (Array (ts, fss))));
+            Plain (Const (I32 (I32.add src 1l) @@ e.at));
+            Plain (Const (I32 (I32.sub sz 1l) @@ e.at));
+            Plain (ArrayCopy (x, y));
+            Refer (Aggr.(ArrayRef (Array (td, fsd))));
+            Plain (Const (I32 dst @@ e.at));
+            Refer (Aggr.(ArrayRef (Array (ts, fss))));
+            Plain (Const (I32 src @@ e.at));
+            Plain (ArrayGet (y, exto));
+            Plain (ArraySet x);
+          ]
+
+      | ArrayFill x, Num (I32 n) :: v :: Num (I32 i) :: Ref (NullRef _) :: vs' ->
+        vs', [Trapping "null array reference" @@ e.at]
+
+      | ArrayFill x, Num (I32 n) :: v :: Num (I32 i) :: Ref Aggr.(ArrayRef (Array (t, fs))) :: vs' ->
+        if I64.gt_u
+             (I64.add (I64_convert.extend_i32_u i) (I64_convert.extend_i32_u n))
+               (I64_convert.extend_i32_u (Lib.List32.length fs)) then
+          vs', [Trapping "out of bounds array access" @@ e.at]
+        else if n = 0l then
+          vs', []
+        else
+          vs', List.map (Lib.Fun.flip (@@) e.at) [
+            Refer (Aggr.(ArrayRef (Array (t, fs))));
+            Plain (Const (I32 i @@ e.at));
+            (e_of_v v e.at);
+            Plain (ArraySet x);
+            Refer (Aggr.(ArrayRef (Array (t, fs))));
+            Plain (Const (I32 (I32.add i 1l) @@ e.at));
+            (e_of_v v e.at);
+            Plain (Const (I32 (I32.sub n 1l) @@ e.at));
+            Plain (ArrayFill x);
+          ]
+
+      | ArrayInitData (x, y), Num (I32 n) :: Num (I32 y_off) :: Num (I32 i) :: Ref (NullRef _) :: vs' ->
+        vs', [Trapping "null array reference" @@ e.at]
+
+      | ArrayInitData (x, y), Num (I32 n) :: Num (I32 y_off) :: Num (I32 i) :: Ref Aggr.(ArrayRef (Array (t, fs))) :: vs' ->
+        if I64.gt_u
+             (I64.add (I64_convert.extend_i32_u i) (I64_convert.extend_i32_u n))
+               (I64_convert.extend_i32_u (Lib.List32.length fs)) then
+          vs', [Trapping "out of bounds array access" @@ e.at]
+        else if data_oob c.frame y y_off n then
+          vs', [Trapping (memory_error e.at Memory.Bounds) @@ e.at]
+        else if n = 0l then
+          vs', []
+        else
+          let ArrayT (FieldT (_mut, st)) = array_type c.frame.inst x in
+          let seg = data c.frame.inst y in
+          let bs = Data.bytes seg in
+          let j = I32.to_int_u y_off in
+          (* TODO: unify with ArrayNewData *)
```

Comment:
Can you extend the Data module with a load_num function analogous to Memory.load_num and use that here and for ArrayNewData?

Reply:
opening Type/Value in Data seems to cause a build circularity. For now I'm factoring this out into a shared helper in `eval.ml`.

Reply:
Hm, that is very odd. What's the circularity? I cannot reproduce it on main.